Source/NSubstitute/Core/ICustomHandlers.cs

```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections.Generic;
+
+namespace NSubstitute.Core
+{
+    /// <summary>
+    ///     Factory method which creates <see cref="ICallHandler" /> from the <see cref="ISubstituteState" />.
+    /// </summary>
+    public delegate ICallHandler CallHandlerFactory(ISubstituteState substituteState);
+
+    public interface ICustomHandlers
+    {
+        IEnumerable<ICallHandler> Handlers { get; }
+        IDictionary<object, object> HandlerDataStorage { get; }
```

Comment:
I think I'd prefer to have each custom call handler manage its own state if that's feasible. 

AutoFixture details aside, I was imagining we could do something like `sub.GetCallRouter().RegisterHandler(new MyHandler(myState))`, so a handler would automatically be associated with a single substitute. Don't know if this is practical, but was my initial line of thinking about this.